File: agentmesh/sdk.py
"""AgentMesh Python SDK: ``@mesh.agent`` decorator, ``mesh.run`` grouping,
and ``mesh.record_usage`` token/cost accounting."""

import contextvars
import functools
import inspect
import time
import uuid
import warnings
from contextlib import contextmanager
from typing import Any, Callable, Dict, Iterator, Optional, TypeVar

from agentmesh.events import Event
from agentmesh.store import EventStore

F = TypeVar("F", bound=Callable[..., Any])

_current_run_id: contextvars.ContextVar[Optional[str]] = contextvars.ContextVar(
    "agentmesh_run_id", default=None
)

_current_span_id: contextvars.ContextVar[Optional[str]] = contextvars.ContextVar(
    "agentmesh_span_id", default=None
)

_current_usage: contextvars.ContextVar[Optional[Dict[str, Any]]] = contextvars.ContextVar(
    "agentmesh_usage", default=None
)


def _new_span_fields() -> Dict[str, str]:
    """Span payload fields for a fresh invocation: its own span_id plus the
    caller's span as ``parent_span_id``, if there is one (AMP v0.2)."""
    fields = {"span_id": str(uuid.uuid4())}
    parent = _current_span_id.get()
    if parent is not None:
        fields["parent_span_id"] = parent
    return fields


class Mesh:
    """Entry point for instrumenting agents.

    If no store is provided, one is resolved per emit so the ``AGENTMESH_DB``
    environment variable is honored even when set after import.
    """

    def __init__(self, store: Optional[EventStore] = None) -> None:
        self._store = store

    def _resolve_store(self) -> EventStore:
        if self._store is not None:
            return self._store
        return EventStore()

    def _emit(self, run_id: str, agent: str, type: str, payload: Dict[str, Any]) -> None:
        self._resolve_store().append(Event.new(run_id, agent, type, payload))

    @contextmanager
    def run(self, run_id: Optional[str] = None) -> Iterator[str]:
        """Group all decorated calls inside this block under one run_id."""
        rid = run_id if run_id is not None else str(uuid.uuid4())
        token = _current_run_id.set(rid)
        try:
            yield rid
        finally:
            _current_run_id.reset(token)

    def current_run_id(self) -> Optional[str]:
        return _current_run_id.get()

    def record_usage(
        self,
        input_tokens: Optional[int] = None,
        output_tokens: Optional[int] = None,
        cost_usd: Optional[float] = None,
        model: Optional[str] = None,
    ) -> None:
        """Accumulate token/cost usage onto the current ``@mesh.agent`` span.

        The accumulated fields are merged into the span's ``agent_end``
        payload (AMP v0.3). Multiple calls within one span sum the token and
        cost numbers; for ``model`` the last non-None value wins. Outside any
        span this is a no-op that emits a warning.
        """
        usage = _current_usage.get()
        if usage is None:
            warnings.warn(
                "mesh.record_usage() called outside any @mesh.agent span; usage dropped",
                stacklevel=2,
            )
            return
        if input_tokens is not None:
            usage["input_tokens"] = usage.get("input_tokens", 0) + input_tokens
        if output_tokens is not None:
            usage["output_tokens"] = usage.get("output_tokens", 0) + output_tokens
        if cost_usd is not None:
            usage["cost_usd"] = usage.get("cost_usd", 0.0) + cost_usd
        if model is not None:
            usage["model"] = model

    def agent(self, name: str) -> Callable[[F], F]:
        """Instrument a sync or async callable as a named agent.

        Emits ``agent_start`` before the call, ``agent_end`` (with
        ``duration_ms``) on success, and ``agent_error`` (with exception
        info) on raise — the exception is re-raised unchanged.

        Each call gets a fresh ``span_id``; nested decorated calls record
        the enclosing call's span as ``parent_span_id`` (AMP v0.2), so a
        trace tree can be rebuilt with ``agentmesh trace RUN_ID``. Usage
        recorded via ``mesh.record_usage`` inside the call is merged into
        the ``agent_end`` payload (AMP v0.3).
        """

        def decorator(fn: F) -> F:
            function_name = getattr(fn, "__qualname__", getattr(fn, "__name__", name))

            def _start(run_id: str, span: Dict[str, str]) -> float:
                payload: Dict[str, Any] = {"function": function_name}
                payload.update(span)
                self._emit(run_id, name, "agent_start", payload)
                return time.perf_counter()

            def _end(
                run_id: str, started: float, span: Dict[str, str], usage: Dict[str, Any]
            ) -> None:
                payload: Dict[str, Any] = {
                    "function": function_name,
                    "duration_ms": (time.perf_counter() - started) * 1000.0,
                }
                payload.update(usage)
                payload.update(span)
                self._emit(run_id, name, "agent_end", payload)

            def _error(
                run_id: str, started: float, exc: BaseException, span: Dict[str, str]
            ) -> None:
                payload: Dict[str, Any] = {
                    "function": function_name,
                    "duration_ms": (time.perf_counter() - started) * 1000.0,
                    "error": str(exc),
                    "error_type": type(exc).__name__,
                }
                payload.update(span)
                self._emit(run_id, name, "agent_error", payload)

            if inspect.iscoroutinefunction(fn):

                @functools.wraps(fn)
                async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                    run_id = _current_run_id.get() or str(uuid.uuid4())
                    span = _new_span_fields()
                    usage: Dict[str, Any] = {}
                    token = _current_span_id.set(span["span_id"])
                    usage_token = _current_usage.set(usage)
                    try:
                        started = _start(run_id, span)
                        try:
                            result = await fn(*args, **kwargs)
                        except BaseException as exc:
                            _error(run_id, started, exc, span)
                            raise
                        _end(run_id, started, span, usage)
                        return result
                    finally:
                        _current_usage.reset(usage_token)
                        _current_span_id.reset(token)

                return async_wrapper  # type: ignore[return-value]

            @functools.wraps(fn)
            def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                run_id = _current_run_id.get() or str(uuid.uuid4())
                span = _new_span_fields()
                usage: Dict[str, Any] = {}
                token = _current_span_id.set(span["span_id"])
                usage_token = _current_usage.set(usage)
                try:
                    started = _start(run_id, span)
                    try:
                        result = fn(*args, **kwargs)
                    except BaseException as exc:
                        _error(run_id, started, exc, span)
                        raise
                    _end(run_id, started, span, usage)
                    return result
                finally:
                    _current_usage.reset(usage_token)
                    _current_span_id.reset(token)

            return sync_wrapper  # type: ignore[return-value]

        return decorator


mesh = Mesh()

File: sdk-ts/src/sdk.ts
/**
 * AgentMesh TypeScript SDK — mirror of the Python `mesh.agent` decorator,
 * `mesh.run` grouping, and `mesh.record_usage` token/cost accounting
 * (agentmesh/sdk.py).
 *
 * Wrapped agents emit `agent_start` before the call, `agent_end` (with
 * `duration_seconds`) on success, and `agent_error` (with exception info,
 * re-thrown unchanged) on failure. Each invocation gets a fresh `span_id`;
 * nested wrapped calls record the enclosing invocation's span as
 * `parent_span_id` (AMP v0.2). Usage recorded via `mesh.recordUsage` inside
 * the call is merged into the `agent_end` payload (AMP v0.3). Context
 * propagates through sync and async calls via AsyncLocalStorage.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import { randomUUID } from "node:crypto";

import { newEvent, type EventType } from "./events.js";
import type { Sink } from "./sink.js";

interface MeshContext {
  runId: string | undefined;
  spanId: string | undefined;
  /** Per-span usage accumulator; present only inside a `mesh.agent` span. */
  usage: Record<string, unknown> | undefined;
}

/** Token/cost usage for `mesh.recordUsage` (camelCase in the API; emitted
 * snake_case on the wire, mirroring Python's keyword arguments). */
export interface UsageRecord {
  inputTokens?: number;
  outputTokens?: number;
  costUsd?: number;
  model?: string;
}

const context = new AsyncLocalStorage<MeshContext>();

function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
  return (
    typeof value === "object" &&
    value !== null &&
    typeof (value as { then?: unknown }).then === "function"
  );
}

function errorPayload(exc: unknown): { error: string; error_type: string } {
  if (exc instanceof Error) {
    return { error: exc.message, error_type: exc.constructor.name };
  }
  return { error: String(exc), error_type: typeof exc };
}

/** Entry point for instrumenting agents. Events go to the given sink. */
export class Mesh {
  private readonly sink: Sink;

  constructor(sink: Sink) {
    this.sink = sink;
  }

  private emit(
    runId: string,
    agent: string,
    type: EventType,
    payload: Record<string, unknown>,
  ): void {
    this.sink.emit(newEvent(runId, agent, type, payload));
  }

  /** Group all wrapped calls inside `fn` under one run_id (mirrors `mesh.run`). */
  run<T>(runId: string | undefined, fn: (runId: string) => T): T {
    const rid = runId ?? randomUUID();
    const parent = context.getStore();
    return context.run({ runId: rid, spanId: parent?.spanId, usage: parent?.usage }, () =>
      fn(rid),
    );
  }

  /** The run_id of the enclosing `mesh.run`, if any. */
  currentRunId(): string | undefined {
    return context.getStore()?.runId;
  }

  /**
   * Accumulate token/cost usage onto the current `mesh.agent` span
   * (mirrors Python's `mesh.record_usage`).
   *
   * The accumulated fields are merged into the span's `agent_end` payload
   * as `input_tokens` / `output_tokens` / `cost_usd` / `model` (AMP v0.3).
   * Multiple calls within one span sum the token and cost numbers; for
   * `model` the last non-null value wins. Nested spans are isolated — a
   * child's usage never leaks onto its parent. Outside any span this is a
   * no-op that emits a `console.warn` per call.
   */
  recordUsage(usage: UsageRecord): void {
    const current = context.getStore()?.usage;
    if (current === undefined) {
      console.warn(
        "mesh.recordUsage() called outside any mesh.agent span; usage dropped",
      );
      return;
    }
    if (usage.inputTokens != null) {
      current["input_tokens"] = ((current["input_tokens"] as number) ?? 0) + usage.inputTokens;
    }
    if (usage.outputTokens != null) {
      current["output_tokens"] =
        ((current["output_tokens"] as number) ?? 0) + usage.outputTokens;
    }
    if (usage.costUsd != null) {
      current["cost_usd"] = ((current["cost_usd"] as number) ?? 0) + usage.costUsd;
    }
    if (usage.model != null) {
      current["model"] = usage.model;
    }
  }

  /**
   * Instrument a sync or async function as a named agent.
   *
   * Calls outside `mesh.run` each get their own fresh run_id, like the
   * Python SDK. For async functions the terminal event is emitted when the
   * returned promise settles.
   */
  agent<Args extends unknown[], R>(name: string, fn: (...args: Args) => R): (...args: Args) => R {
    const functionName = fn.name !== "" ? fn.name : name;
    const mesh = this;
    return function wrapped(...args: Args): R {
      const enclosing = context.getStore();
      const runId = enclosing?.runId ?? randomUUID();
      const spanId = randomUUID();
      const span: Record<string, string> = { span_id: spanId };
      if (enclosing?.spanId !== undefined) {
        span["parent_span_id"] = enclosing.spanId;
      }
      const usage: Record<string, unknown> = {};
      const started = Date.now();
      const end = (): void => {
        mesh.emit(runId, name, "agent_end", {
          function: functionName,
          duration_seconds: (Date.now() - started) / 1000,
          ...usage,
          ...span,
        });
      };
      const error = (exc: unknown): void => {
        mesh.emit(runId, name, "agent_error", {
          function: functionName,
          duration_seconds: (Date.now() - started) / 1000,
          ...errorPayload(exc),
          ...span,
        });
      };

      mesh.emit(runId, name, "agent_start", { function: functionName, ...span });
      let result: R;
      try {
        result = context.run({ runId, spanId, usage }, () => fn(...args));
      } catch (exc) {
        error(exc);
        throw exc;
      }
      if (isPromiseLike(result)) {
        return (result as PromiseLike<unknown>).then(
          (value) => {
            end();
            return value;
          },
          (exc: unknown) => {
            error(exc);
            throw exc;
          },
        ) as R;
      }
      end();
      return result;
    };
  }
}
